packages/webpack-cli/lib/webpack-cli.js

```diff
@@ -185,6 +187,250 @@ class WebpackCLI {
+            }).reduce((a, i) => a.concat(i), []);
+        };
+
+        const getConfigInfoFromFileName = (filename) => {
+            const ext = extname(filename);
+            // since we support only one config for now
+            const allFiles = [filename];
+            // return all the file metadata
+            return allFiles
+                .map((file) => {
+                    return {
+                        path: resolve(file),
+                        ext: ext,
+                        module: extensions[ext] || null,
+                    };
+                })
+                .filter((e) => existsSync(e.path));
+        };
+
+        // Reads a config file given the config metadata
+        const requireConfig = (configModule) => {
+            const extension = Object.keys(jsVariants).find((t) => configModule.ext.endsWith(t));
+
+            if (extension) {
+                rechoir.prepare(extensions, configModule.path, process.cwd());
+            }
+
+            let config = require(configModule.path);
+
+            if (config.default) {
+                config = config.default;
+            }
+
+            return { config, path: configModule.path };
+        };
+
+        // Responsible for reading user configuration files
+        // else does a default config lookup and resolves it.
+        const resolveConfigFiles = async (args) => {
+            const { config, mode } = args;
+
+            if (config && config.length > 0) {
+                const resolvedOptions = [];
+                const finalizedConfigs = config.map(async (webpackConfig) => {
+                    const configPath = resolve(webpackConfig);
+                    const configFiles = getConfigInfoFromFileName(configPath);
+
+                    if (!configFiles.length) {
+                        logger.error(`The specified config file doesn't exist in ${configPath}`);
+                        process.exit(2);
+                    }
+
+                    const foundConfig = configFiles[0];
+                    const resolvedConfig = requireConfig(foundConfig);
+
+                    return finalize(resolvedConfig, args);
+                });
+
+                // resolve all the configs
+                for await (const resolvedOption of finalizedConfigs) {
+                    if (Array.isArray(resolvedOption.options)) {
+                        resolvedOptions.push(...resolvedOption.options);
+                    } else {
+                        resolvedOptions.push(resolvedOption.options);
+                    }
+                }
+
+                opts['options'] = resolvedOptions.length > 1 ? resolvedOptions : resolvedOptions[0] || {};
+
+                return;
+            }
+
+            // When no config is supplied, lookup for default configs
+            const defaultConfigFiles = getDefaultConfigFiles();
+            const tmpConfigFiles = defaultConfigFiles.filter((file) => {
+                return existsSync(file.path);
```

Comment:
We should avoid using `sync` command here, it is slow, let's migrate on async API

Reply:
let's do it in refactor